integration-tests/tests/src/tests/counter.ts

```diff
@@ -0,0 +1,994 @@
+        });
+        expect(object.nullableCounter).to.be.null;
+
+        this.realm.write(() => {
+          // @ts-expect-error Cannot currently express in TS that a Counter can be set to a number (while 'get' returns Counter).
+          object.nullableCounter = 1;
+        });
+        expectCounter(object.nullableCounter);
+        expect(object.nullableCounter.value).equals(1);
+
+        this.realm.write(() => {
+          object.nullableCounter = undefined;
+        });
+        expect(object.nullableCounter).to.be.null;
+
+        this.realm.write(() => {
+          // @ts-expect-error Cannot currently express in TS that a Counter can be set to a number (while 'get' returns Counter).
+          object.nullableCounter = -100_000;
+        });
+        expectCounter(object.nullableCounter);
+        expect(object.nullableCounter.value).equals(-100_000);
+      });
+
+      for (const updateMode of [UpdateMode.Modified, UpdateMode.All]) {
+        it(`updates nullable counter from int -> null -> int via UpdateMode: ${updateMode}`, function (this: RealmContext) {
+          const object = this.realm.write(() => {
+            return this.realm.create<IWithNullableCounter>(WithNullableCounterSchema.name, {
+              nullableCounter: 0,
+            });
+          });
+          expectCounter(object.nullableCounter);
+
+          const _id = object._id;
+
+          this.realm.write(() => {
+            this.realm.create<IWithNullableCounter>(
+              WithNullableCounterSchema.name,
+              {
+                _id,
+                nullableCounter: null,
+              },
+              updateMode,
+            );
+          });
+          expect(object.nullableCounter).to.be.null;
+
+          this.realm.write(() => {
+            this.realm.create<IWithNullableCounter>(
+              WithNullableCounterSchema.name,
+              {
+                _id,
+                nullableCounter: 1,
+              },
+              updateMode,
+            );
+          });
+          expectCounter(object.nullableCounter);
+          expect(object.nullableCounter.value).equals(1);
+
+          this.realm.write(() => {
+            this.realm.create<IWithNullableCounter>(
+              WithNullableCounterSchema.name,
+              {
+                _id,
+                nullableCounter: null,
+              },
+              updateMode,
+            );
+          });
+          expect(object.nullableCounter).to.be.null;
+
+          this.realm.write(() => {
+            this.realm.create<IWithNullableCounter>(
+              WithNullableCounterSchema.name,
+              {
+                _id,
+                nullableCounter: -100_000,
+              },
+              updateMode,
+            );
+          });
+          expectCounter(object.nullableCounter);
+          expect(object.nullableCounter.value).equals(-100_000);
+        });
+      }
+    });
+  });
+
+  describe("Filtering", () => {
```

Comment:
While I understand why this goes here, it's also interesting in terms of how we structure our tests going forward as it could also have been an addendum to https://github.com/realm/realm-js/blob/main/integration-tests/tests/src/tests/queries.ts

Reply:
Yeah I think for the counters tests it still made sense to keep most things as it's easier to remember what's allowed in terms of counters. But agree that it would be valuable for us to decide on a clear structure of our tests, where it's easy to quickly reason about where a new test should go are where a current test is located.